RuntimeLibcalls: Split lowering decisions into LibcallLoweringInfo

arsenm opened the pull request:
Introduce a new class for the TargetLowering usage. This tracks the
subtarget specific lowering decisions for which libcall to use.
RuntimeLibcallsInfo is a module level property, which may have multiple
implementations of a particular libcall available. This attempts to be
a minimum boilerplate patch to introduce the new concept.

In the future we should have a tablegen way of selecting which
implementations should be used for a subtarget. Currently we
do have some conflicting implementations added, it just happens
to work out that the default cases to prefer is alphabetically
first (plus some of these still are using manual overrides
in TargetLowering constructors).

arsenm committed:
RuntimeLibcalls: Split lowering decisions into LibcallLoweringInfo

Introduce a new class for the TargetLowering usage. This tracks the
subtarget specific lowering decisions for which libcall to use.
RuntimeLibcallsInfo is a module level property, which may have multiple
implementations of a particular libcall available. This attempts to be
a minimum boilerplate patch to introduce the new concept.

In the future we should have a tablegen way of selecting which
implementations should be used for a subtarget. Currently we
do have some conflicting implementations added, it just happens
to work out that the default cases to prefer is alphabetically
first (plus some of these still are using manual overrides
in TargetLowering constructors).

compnerd commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -193,6 +193,58 @@ struct MemOp {
+  // FIXME: Remove this wrapper in favor of directly using
+  // getLibcallImplCallingConv
+  LLVM_ABI CallingConv::ID getLibcallCallingConv(RTLIB::Libcall Call) const {
+    return RTLCI.LibcallImplCallingConvs[LibcallImpls[Call]];
```
Mind changing this to:

```
  return RTLCI.LibcallImplCallingConvs[getLibcallImpl(Call)];
```

arsenm replied:
This is moving code, I don't want to mix other style changes in

compnerd commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -193,6 +193,58 @@ struct MemOp {
+      return RTLIB::RuntimeLibcallsInfo::getLibcallImplName(Memcpy);
+
+    // Fallback to memmove if memcpy isn't available.
+    return getLibcallName(RTLIB::MEMMOVE);
```
I would invert this:

```
  if (Memcpy == RTLIB::Unsupported)
    return getLibcallName(RTLIB::MEMMOVE);
  return RTLIB::RuntimeLibcallsInfo::getLibcallImplName(Memcpy);
```

arsenm replied:
This was just a code move

arsenm commented on llvm/include/llvm/CodeGen/TargetLowering.h:
```diff
@@ -193,6 +193,58 @@ struct MemOp {
+  /// Get the libcall routine name for the specified libcall.
+  // FIXME: This should be removed. Only LibcallImpl should have a name.
+  LLVM_ABI const char *getLibcallName(RTLIB::Libcall Call) const {
+    // FIXME: Return StringRef
```
But the main issue is all the places relying on null checks of this. It's more a matter of migrating to direct supported checks instead of string based logic

arsenm committed:
CodeGen: Move LibcallLoweringInfo to be with RuntimeLibcallInfo

Eventually this will be program dependent state, so it should not
be part of TargetLowering. It's easiest to have this as part of
the analysis that will provide RuntimeLibcallInfo. In principle
we would also need this to more precisely refine the set of libcalls
that full LTO could trim based on the subtarget.

arsenm committed:
CodeGen: Move LibcallLoweringInfo to separate file

arsenm commented:
ping

compnerd commented on llvm/include/llvm/CodeGen/LibcallLoweringInfo.h:
```diff
@@ -0,0 +1,65 @@
+      return RTLIB::RuntimeLibcallsInfo::getLibcallImplName(Memcpy);
+
+    // Fallback to memmove if memcpy isn't available.
+    return getLibcallName(RTLIB::MEMMOVE);
```
Inverting this would be more LLVM style:


```suggestion
    RTLIB::LibcallImpl Memcpy = getLibcallImpl(RTLIB::MEMCPY);
    if (Memcpy == RTLIB::Unsupported)
      // Fallback to memmove if memcpy isn't available.
      return getLibcallName(RTLIB::MEMMOVE);
    return RTLIB::RuntimeLibcallsInfo::getLibcallImplName(Memcpy);


```

arsenm committed:
Invert condition

arsenm committed:
update test

arsenm committed:
Fix merge
